src/Sound/SoundHandler.cpp

```diff
@@ -0,0 +1,182 @@
+ * Copyright (c) 2018-2020 openblack developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "SoundHandler.h"
+
+#include "3D/Camera.h"
+#include <Game.h>
+#include <fstream>
+#include <vector>
+
+namespace openblack::audio
+{
+
+SoundHandler::SoundHandler(std::unique_ptr<AudioDecoder>&& decoder, std::unique_ptr<AudioPlayer>&& player) : _audioDecoder(std::move(decoder)), _audioPlayer(std::move(player))
+{
+	_emitterId = 0;
+}
+
+void SoundHandler::RegisterSoundPack(std::unique_ptr<SoundPack>& soundPack)
+{
+	spdlog::debug("Registering SoundPack {}", soundPack->GetFileName());
+	_soundPacks.push_back(std::move(soundPack));
+	auto last = _soundPacks.back();
+	for (auto& [id, sound] : last->GetSounds())
+	{
+		if (_soundIdLookup.find(id) == _soundIdLookup.end())
+			_soundIdLookup[id] = last;
+		else
+			spdlog::warn("Duplicate sound ID (\"{}\"). Unable to add sound", id);
+	}
+}
+
+void SoundHandler::Tick(Game& game)
+{
+	// Update listener state
+	auto& cam = game.GetCamera();
+	auto pos = cam.GetPosition();
+	auto vel = cam.GetVelocity();
+	auto forward = cam.GetForward();
+	auto top = cam.GetUp();
+	_audioPlayer->UpdateListenerState(pos, vel, forward, top);
+
+	// Update active user emitter states
+	for (auto& [emitterId, emitter] : _emitters) _audioPlayer->UpdateEmitterState(emitter);
+
+	// Sort activated emitters
+	// TODO: Implement
+
+	// Release hardware resources
+	// Update emitters with hardware resources
+	// Assign hardware resources
+//	for (auto& [id, emitter] : _emitters)
+//	{
+//		_audioPlayer->CleanUpResource(emitter);
+//	}
+
+	// Cull stopped emitters
+	for (auto iterator = _emitters.begin(); iterator != _emitters.end();)
+	{
+		auto& [id, emitter] = *iterator;
+		auto audioStatus = _audioPlayer->GetAudioStatus(emitter.audioSourceId);
+
+		if (emitter.loop == PlayType::Once && audioStatus == AudioStatus::Stopped)
+		{
+			_audioPlayer->CleanUpResource(emitter);
+			iterator = _emitters.erase(iterator);
+		}
+		else
+			iterator++;
+	}
+}
+
+std::unique_ptr<Sound>& SoundHandler::GetSound(SoundId id)
+{
+	auto& sound = _soundIdLookup[id]->GetSound(id);
+
+	if (!sound->loaded)
+	{
+		_audioDecoder->ToPCM16(*sound);
+		sound->loaded = true;
+	}
+
+	return sound;
+}
+
+const std::map<AudioEmitterId, AudioEmitter>& SoundHandler::GetEmitters() const
+{
+	return _emitters;
+}
+
+AudioEmitterId SoundHandler::CreateEmitters(const std::vector<SoundId> ids, glm::vec3& position, glm::vec3& velocity, glm::vec3& direction, glm::vec2& radius, float volume, bool world, PlayType playType)
+{
+	std::vector<SoundId> out;
+	std::sample(ids.begin(), ids.end(), std::back_inserter(out), 1, std::mt19937{std::random_device{}()});
```

Comment:
This random device should live in the class instead of always creating it.
Also, we should have a randomization service to centralize things.
This is also needed for AI.

---

Also, why not use this:
```c++
assert(ids.size > 0);
std::default_random_engine generator; // Need to seed and move to class or service
std::uniform_int_distribution<size_t> distribution(0, ids.size() - 1);
SoundId out = ids[distribution(generator)];
```
Which could obviously be abstracted within a random service to be something like
```c++
template<T>
const T& RandomService::PickOne(const std::vector<T> container, size_t minIdx = 0, size_t maxIdx = -1)
{
    assert(container.size() > 0);
    assert(minIdx < ids.size());
    if (maxIdx == -1)
    {
        maxIdx = ids.size() - 1;
    }
    assert(maxIdx < ids.size());
    std::uniform_int_distribution<size_t> distribution(minIdx, maxIdx);
    return container[distribution(m_generator)];
}
```